### PR TITLE
MGMT-13203: Create REST points for validation ignore feature.

### PR DESCRIPTION
**Pull request by paul-maidment**
[MGMT-13203](https://issues.redhat.com//browse/MGMT-13203): Create REST points for validation ignore feature.

This provides the API endpoint for this feature enhancement https://issues.redhat.com/browse/MGMT-12547

The aim of the feature is to allow the customer to ignore validations if they have org permissions to do so.
This PR implements the API enpoint to allow the customer to ignore the validations.

We have decided to keep this separate from the cluster API as we are not actively encouraging use of this feature.

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by paul-maidment** [comment]
@cc @avishayt

**Comment by paul-maidment** [comment]
cc: @avishayt

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3501]:
```suggestion
  /v2/clusters/{cluster_id}/ignored-validations:
```

Let's save an API and just specify both cluster and host validations in the body - one JSON for each.
Also, need a GET API to return the current value.

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3510]:
```suggestion
          description: The cluster whose failing validations should be ignored according to this list.
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3527]:
```suggestion
          description: The cluster whose failing validations should be ignored according to this list
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3533]:
```suggestion
          description: The validations to be ignored.
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3518]: There's no reason to have another API to return clusters - let's remove this GET API.

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3567]:
```suggestion
      cluster-validation-ids:
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:3571]:
```suggestion
      host-validation-ids:
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:4934]:
```suggestion
        description: JSON-formatted list of host validation IDs that will be ignored.  It may also contain a list with a single string "all" to ignore all host validations.  Some validations cannot be ignored.
```

File: swagger.yaml
Comment (avishayt) [inline swagger.yaml:4938]:
```suggestion
        description: JSON-formatted list of cluster validation IDs that will be ignored.  It may also contain a list with a single string "all" to ignore all cluster validations.  Some validations cannot be ignored.
```

**Comment by avishayt** [comment]
Looks good to me. Not sure about the split of code between inventory.go and inventory_v2_handlers.go. Let's have another pair of eyes on this.

File: swagger.yaml
Comment (omertuc) [inline swagger.yaml:3572]: I don't like that the host and cluster ignored validations are bundled up like this, clusters are one entity, hosts are another entity, each of them should have its own field.

Mostly because hosts are not always bound to clusters and can move between them

Comment (paul-maidment) [reply swagger.yaml:3572]: This is a cluster level setting and should not be applied on a "per host" basis, that's not the intention here.

The intention is that for very specific customers, we would like to enable them to mark specific validations as skippable across the board. What this setting means is "during the installation of this cluster, whenever you encounter this host validation, ignore it..."

If we go the other way, registering this on a per host basis, it's possible but it would complicate the usage of this feature for the customer. 

Open to change my mind if there is broader support for changing in this way.

Comment (omertuc) [reply swagger.yaml:3572]: Hosts have validations even when detached from clusters, how would users skip those then?

Comment (omertuc) [reply swagger.yaml:3572]: At least, they do as far as I know. I may be wrong about this

Comment (paul-maidment) [reply swagger.yaml:3572]: Checking this point, I didn't make it clear in the enhancement doc but unbound clusters are a non-goal for this feature. This is really for those customers who have a "more or less" viable cluster ready to install but are blocked by one validation that would allow the installation to proceed if they could just skip it.

Comment (avishayt) [reply swagger.yaml:3572]: I'm sorry I missed this earlier, but @omertuc has a good point.  We may want to expose this functionality in our Kubernetes API, and in late binding, we do have some host validations that run before binding.  Worse, we block binding if a host has failing validations.

Comment (filanov) [reply swagger.yaml:3572]: kube-api is one thing late binding is another thing, after all the controller will call what ever it needs. 
the downside of having it in the host level is that users will need to do it per host, on the other had having it in the infra-env level will mean that users that use late binding (we don't have many of those as far as i know) will affect all the hosts in the infra-env. 
So i assume that having the field for each host is more flexible. but it's sounds like an overkill for something that we don't want to officially support. 
for me for now it's enough to have it on the cluster level and if we see a demand we can decide on how to improve, it's not an api that we need to maintain and not break.

Comment (omertuc) [reply swagger.yaml:3572]: > that users will need to do it per host

Not necessarily, we could have a hybrid solution where hosts inherit the validation-skips from their cluster, and also have their own list

Comment (omertuc) [reply swagger.yaml:3572]: A hybrid solution can also be done in parts, i.e. first merge this PR as is, then add per-host skips in the future

Comment (omertuc) [reply swagger.yaml:3572]: The downside of a hybrid solution is that it's much more complex in the backend - two tiers of validation skips that have to be merged

Comment (paul-maidment) [reply swagger.yaml:3572]: The suggestion from @omertuc sounds reasonable as part of a strategy where we see whether or not there is demand to expand the functionality further.

Comment (paul-maidment) [reply swagger.yaml:3572]: "we will need to think how to add it to kube-api without having that directly in the spec because we don't want to expose it as part of the official api" -> What about a custom CRD that accepts a cluster reference?

Comment (filanov) [reply swagger.yaml:3572]: i was thinking about annotations, but for now i would not even add it to kube-api at this point.
reference is still something that you will have in the spec.

Comment (paul-maidment) [reply swagger.yaml:3572]: So for now, this REST API should do the job?

Comment (paul-maidment) [reply swagger.yaml:3572]: So for now, this REST API should do the job?

Comment (filanov) [reply swagger.yaml:3572]: yes, lets start with that and see if it's used.

Comment (avishayt) [reply swagger.yaml:3572]: > improve only on demand

It will be harder to change later once it's in the DB.  Why not fix it now?

Comment (omertuc) [reply swagger.yaml:3572]: > > improve only on demand
> 
> It will be harder to change later once it's in the DB. Why not fix it now?

Assuming we're going for an eventual hybrid solution, you still need the clusters to specify both host and cluster validations, so clusters can affect all of their hosts in bulk, making it simpler for users

Comment (omertuc) [reply swagger.yaml:3572]: > I think a hybrid solution would be even more complicated because we would need to deal with different values on the host vs on the cluster.

Yes, that's the point I was making [here](https://github.com/openshift/assisted-service/pull/4927#discussion_r1094429262)

Although the combination is a simple union(set(host_skips) | set(cluster_host_skips))

Comment (paul-maidment) [reply swagger.yaml:3572]: @filanov looking for your further input here...

Comment (filanov) [reply swagger.yaml:3572]: if it will be on the host level we will need to handle cases where host is reused, isn't that an overkill? 
Lets remember the goal, it's to skip validation in a very specific cases to enable starting instlalation, and it's not supported api. 
late binding can be a non goal, if the infra-env have a cluster reference host will not be unbound so the validations are not relevant there. 

> It will be harder to change later once it's in the DB. Why not fix it now?
what is the concern about the DB? cluster will have a list regardless which solution we have.

Comment (avishayt) [reply swagger.yaml:3572]: > what is the concern about the DB? cluster will have a list regardless which solution we have.
Either the cluster has two lists (one for cluster and one for host validations), o the cluster resource has one list and the host resource has one list.

What about a host with some hardware that we're not validating properly, or someone wants to install it with a smaller disk?  The validation should be ignored regardless of the cluster that the host is bound to.

Comment (filanov) [reply swagger.yaml:3572]: Maybe i'll ask it in a different way, with the existing solution besides late binding, do we have any case that will prevent the user to start installation if he is using the api that suggested here? and the second question is, what in this solution blocking us from making improvements later (Avishay said DB, not sure i understand why)?

Comment (avishayt) [reply swagger.yaml:3572]: I think it will be relatively little effort to change now.  Later it will be more difficult.  We'll add it to the host, then people will ask why it's both on the host and on the cluster, then we'll remove it from the cluster.

Comment (paul-maidment) [reply swagger.yaml:3572]: another option is to dispose of this whole host vs cluster flexibility, make it a single flag "ignore all validations (with the exception of the ones we can't ignore)"

So I could get rid of the JSON fields and replace with a boolean.

@ori-amizur made this suggestion, also sounds reasonable?

Comment (filanov) [reply swagger.yaml:3572]: > I think it will be relatively little effort to change now. Later it will be more difficult. We'll add it to the host, then people will ask why it's both on the host and on the cluster, then we'll remove it from the cluster.

OK lets go with that then. 
two different apis, one for host and one for cluster. in hat case we have it per host so users will have a bit more work but ti's fine because the api is not recommended anyway

Comment (paul-maidment) [reply swagger.yaml:3572]: @filanov before we go ahead, should we consider the suggestion from @ori-amizur ?

Comment (paul-maidment) [reply swagger.yaml:3572]: Then let's just stick with what we have in this PR and not overcomplicate things

Comment (filanov) [reply swagger.yaml:3572]: yes please

Comment (filanov) [reply swagger.yaml:3572]: Resolving this conversation. 
The current api will provide a good solution for all the users, we are ignoring late binding for this api. 
as described it's a patch solution for blocked users and not a supported api.

**Comment by paul-maidment** [comment]
Description updated

File: swagger.yaml
Comment (omertuc) [inline swagger.yaml:3572]: shouldn't the `cluster:` definition have `ignored-validations` under it?

File: swagger.yaml
Comment (omertuc) [inline swagger.yaml:3582]: I think this should explicitly mention "... that will be ignored for all hosts that belong to this cluster"

File: swagger.yaml
Comment (omertuc) [inline swagger.yaml:3578]:
```suggestion
        description: JSON-formatted list of cluster validation IDs that will be ignored.  It may also contain a list with a single string "all" to ignore all cluster validations. Some validations cannot be ignored.
```

**Comment by omertuc** [comment]
We should have a feature usage flag for this feature, so we can easily filter out clusters that have it

**Comment by omertuc** [comment]
What is the dedicated GET endpoint for? Why not just return it along with all cluster fields when the user GETs a cluster as we do for all other cluster fields?

**Comment by omertuc** [comment]
Also we already have a PATCH API for clusters to change cluster fields, we shouldn't have a dedicated endpoint for validations either

**Comment by paul-maidment** [comment]
> We should have a feature usage flag for this feature, so we can easily filter out clusters that have it

We do, but it's part of another ticket in the epic, the feature usage is being recorded at the moment the cluster transitions to installing is this is the moment at which we can say this feature was used.

**Comment by paul-maidment** [comment]
> What is the dedicated GET endpoint for? Why not just include those fields among all cluster fields when the user GETs an entire cluster as we do for all other cluster fields?

This was previously discussed with @avishayt and @filanov the main reason is to keep this isolated from existing API's as it's not a feature we are actively encouraging use of.

File: swagger.yaml
Comment (omertuc) [reply swagger.yaml:3572]: We have a `cluster:` [object](https://github.com/openshift/assisted-service/blob/2688136ed6398480e15eb6575466c265f92983e2/swagger.yaml#L4688), the ignored host validations and ignored cluster validations should be a parameter in that object

**Comment by paul-maidment** [comment]
> Also we already have a PATCH API for clusters to change cluster fields, we shouldn't have a dedicated endpoint for ignoring validations

See previous comment

**Comment by omertuc** [comment]
> > What is the dedicated GET endpoint for? Why not just include those fields among all cluster fields when the user GETs an entire cluster as we do for all other cluster fields?
> 
> This was previously discussed with @avishayt and @filanov the main reason is to keep this isolated from existing API's as it's not a feature we are actively encouraging use of.

Ack

File: swagger.yaml
Comment (paul-maidment) [reply swagger.yaml:3572]: Sorry, I disagree, we made a purposeful decision to exclude this from the cluster api

Comment (filanov) [reply swagger.yaml:3572]: the suggestion from @avishayt was not to include it in the cluster object because it's not a supported api

Comment (omertuc) [reply swagger.yaml:3572]: Alright. Next time please mention such non-obvious decisions in the commit message

Comment (paul-maidment) [reply swagger.yaml:3572]: That's a fair point, I'll amend the commit message now.

File: internal/common/validations.go
Comment (omertuc) [inline internal/common/validations.go:27]: Why is this not a simple function? Why var?

Comment (paul-maidment) [reply internal/common/validations.go]: Good point, just fixed that.

File: internal/common/validations.go
Comment (omertuc) [reply internal/common/validations.go:20]: And this one has no callers outside the package, it should be privat

File: internal/common/validations.go
Comment (omertuc) [inline internal/common/validations.go:39]: `strings.Join` is for the sole purpose of displaying something to the user, which this function should not be responsible for. It's very confusing to see the join here as there's no obvious reason for it to happen in this context. The caller should perform the join instead, this function should return a simple array

**Comment by omertuc** [comment]
[This](https://github.com/openshift/assisted-service/pull/4927#discussion_r1094817565) comment is in a resolved conversation but it was not resolved

File: internal/common/validations.go
Comment (paul-maidment) [reply internal/common/validations.go]: The private thing..it will have callers outside of the package when the next PR is applied, so if I make it private now, I have to change it later.

Comment (paul-maidment) [reply internal/common/validations.go]: > [This](https://github.com/openshift/assisted-service/pull/4927#discussion_r1094817565) comment is in a resolved conversation but it was not resolved
I have unresolved so that we can discuss, we can make this private, only to make it public in the next PR, seems fruitless to me.

Comment (omertuc) [reply internal/common/validations.go]: ok

File: internal/common/validations.go
Comment (paul-maidment) [reply internal/common/validations.go:39]: Done, join operation moved to the caller and changed return a string array

**Comment by paul-maidment** [comment]
@omertuc I have addressed the review points made and would like to request further feedback or an LGTM

**Review by filanov — APPROVED** [review]

**Review by filanov — APPROVED** [review]

**Pull request merged**
